### PR TITLE
Make retries configurable and exponential. Fixed a bug with incrementing retries

### PR DESCRIPTION
**Pull request by rajesh-blueshift**
### Description
- Fixed a bug with retry logic. Retries were incremented twice
- Made retries exponential
- Increased number of retries
- Made retries configurable at an App level (to do some more testing later without needing to change code)

### Jira Tickets (IDs in brackets will be auto-linked)
- [CD-2186] - [YesPark] High bounce rate for push messages campaigns

### Test Cases
- [x] Check out this branch locally
- [x] Use the path to this branch in the `bsft` Gemfile
- [x] Test a push campaign for regression
- [x]  Comment out the code inside the function `ok` - lines 95 and 96. Try a test send. Verify that the notification retries and fails eventually. (Note: You will get notifications on your device while you test this)
- [x] Manually hardcode the return code to something like 503 in the `handle_response` function to emulate a service error.  Verify that the notification retries and fails eventually. (Note: You will get notifications on your device while you test this)

[CD-2186]: https://blueshift.atlassian.net/browse/CD-2186?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/rpush/daemon/dispatcher/apns_http2.rb [hunk 1]
```diff
@@ -43,7 +43,14 @@ def create_http2_client(app)
           url = URLS[app.environment.to_sym]
           client = NetHttp2::Client.new(url, ssl_context: prepare_ssl_context, connect_timeout: DEFAULT_TIMEOUT)
           client.on(:error) do |error|
-            log_error(error)
+            if error.is_a?(Errno::ECONNRESET)
+              log_warn(
```
Comment (rajesh-blueshift) [inline lib/rpush/daemon/dispatcher/apns_http2.rb:47]: This is to avoid frequent back traces in the rpush log

File: lib/rpush/daemon/batch.rb [hunk 2]
```diff
@@ -27,12 +27,18 @@ def each_delivered(&blk)
         @delivered.each(&blk)
       end
 
-      def mark_retryable(notification, deliver_after)
+      def mark_retryable(notification, deliver_after, update_store=true)
```
Comment (chongbomb) [inline lib/rpush/daemon/batch.rb]:
```suggestion
      def mark_retryable(notification, deliver_after, update_store = true)
```

File: lib/rpush/daemon/apns2/delivery.rb [hunk 3]
```diff
@@ -26,20 +26,19 @@ def perform
           # If there are notifications that are not yet delivered
           # Usually means that there was a Connection reset error
           # Mark remaining notifications for bounded retry
-          mark_batch_retryable_with_max_retries(
-            Time.now + 10.seconds,
+          mark_batch_retryable_exponential(
```
Comment (chongbomb) [inline lib/rpush/daemon/apns2/delivery.rb:29]: What about the batch retry in the service_unavailable method?

**Review by chongbomb — COMMENTED** [review]
How did you test that retries worked?

File: lib/rpush/daemon/apns2/delivery.rb (hunk 3 above)
Comment (rajesh-blueshift) [reply lib/rpush/daemon/apns2/delivery.rb:29]: Yeah I think I missed that

Comment (rajesh-blueshift) [reply lib/rpush/daemon/apns2/delivery.rb:29]: Fixed this now

File: lib/rpush/daemon/batch.rb (hunk 2 above)
Comment (rajesh-blueshift) [reply lib/rpush/daemon/batch.rb]: Sure, done

**Comment by rajesh-blueshift** [comment]
> How did you test that retries worked?

I tested using the new steps added in this PR's description. You can try those steps as well to test

**Review by chongbomb — APPROVED** [review]

**Pull request merged**